code/modules/mob/abstract/new_player/new_player.dm

```diff
@@ -295,50 +296,12 @@ INITIALIZE_IMMEDIATE(/mob/abstract/new_player)
 		global_announcer.autosay("A new[rank ? " [rank]" : " visitor" ] [join_message ? join_message : "has arrived on the station"].", "Arrivals Announcement Computer")
 
 /mob/abstract/new_player/proc/LateChoices()
-	var/name = client.prefs.real_name
-
-	var/dat = "<center>"
-	dat += "<b>Welcome, [name].<br></b>"
-	dat += "Round Duration: [get_round_duration_formatted()]<br>"
-	dat += "Alert Level: [capitalize(get_security_level())]<br>"
-
-	if(emergency_shuttle) //In case Nanotrasen decides reposess CentComm's shuttles.
-		if(emergency_shuttle.going_to_centcom()) //Shuttle is going to centcomm, not recalled
-			dat += "<span class='warning'><b>The station has been evacuated.</b></span><br>"
-		if(emergency_shuttle.online())
-			if (emergency_shuttle.evac)	// Emergency shuttle is past the point of no recall
-				dat += "<span class='warning'>The station is currently undergoing evacuation procedures.</span><br>"
-			else						// Crew transfer initiated
-				dat += "<span class='warning'>The station is currently undergoing crew transfer procedures.</span><br>"
-
-	var/unique_role_available = FALSE
-	for(var/ghost_role in SSghostroles.spawners)
-		var/datum/ghostspawner/G = SSghostroles.spawners[ghost_role]
-		if(!G.show_on_job_select)
-			continue
-		if(G.cant_see(src))
-			continue
-		unique_role_available = TRUE
-		break
-
-	if(unique_role_available)
-		dat += "<font color='[COLOR_BRIGHT_GREEN]'><b>A unique ghost role is available:</b></font><br>"
-	dat += "<a href='byond://?src=\ref[src];ghostspawner=1'>Ghost Spawner Menu</A><br>"
-
-	dat += "Choose from the following open/valid positions:<br>"
-	for(var/datum/job/job in SSjobs.occupations)
-		if(job && IsJobAvailable(job.title))
-			var/active = 0
-			// Only players with the job assigned and AFK for less than 10 minutes count as active
-			for(var/mob/M in player_list) //Added isliving check here, so it won't check ghosts and qualify them as active
-				if(isliving(M) && M.mind && M.client && M.mind.assigned_role == job.title && M.client.inactivity <= 10 MINUTES)
-					active++
-			dat += "<a href='byond://?src=\ref[src];SelectedJob=[job.title]'>[client.prefs.GetPlayerAltTitle(job)] ([job.current_positions]) (Active: [active])</a><br>"
-
-	dat += "</center>"
-	send_theme_resources(src)
-	src << browse(enable_ui_theme(src, dat), "window=latechoices;size=300x640;can_close=1")
-
+	if(istype(late_choices_ui)) // refresh the UI
+		late_choices_ui.ui.check_for_change()
+	else
+		// This creates the latechoices UI.
+		// It is automatically created and then destroyed and unassigned when closed.
+		late_choices_ui = new(src)
```

Comment:
Since these are per-player is it not possible to cache the gFI icon on the datum?

Reply:
There's no difference, the datum disappears as soon as the dialogue closes.

And the reason why I made it refresh on (re-)opening is so that you can change the character in character loadout and see the change.

_Maybe_ it could be done so that the UI refreshes all the time (to see character, alert level and whatnot changes immediately) _but_ cache the icon and update it only when the character actually changes. But that would still update it as much as now; we'd just get the benefits of the constantly updating UI.